New Android TV media player integration

a1ex4 opened the pull request:
## Description:
This component allows to use any Android TV / Android device as a media player in Home-Assistant.

Right now there are two ways to communicate with an Android device, through the ADB protocol: use the Python implementation of the protocol, or connect to an external ADB server that is connected to the Android device. Given the many different combinations of Android devices and machines hosting HA, both methods are implemented in this component, trying to make the experience as streamlined as possible.

This PR is no longer WIP 🙂 
Thanks in advance for any input! 

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#7788

## Example entry for `configuration.yaml` (if applicable):
```yaml
media_player:
  # a device that does not require ADB authentication, or uses the generated ADB keys's default location
  - platform: androidtv
    host: 192.168.1.37
    name: MIBOX3
      
  # a device using moved ADB keys
  - platform: androidtv
    host: 192.168.1.37
    name: MIBOX3
    adbkey: /config/adbkey

  # using an external ADB server to connect to the device
  - platform: androidtv
    host: 192.168.1.37
    name: MIBOX3
    adb_server_ip: 127.0.0.1

  # custom user defined "known apps"
  - platform: androidtv
    host: 192.168.1.37
    name: MIBOX3
    apps:
      "amazon": "Amazon Premium Video"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

a1ex4 committed:
Added androidtv.py

marchingphoenix commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,481 @@
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['https://github.com/JeffLIrion/python-androidtv/zipball/pure-python-adb#androidtv==0.0.3']
```
I don't think you can add requirements this way. I'll have to check.

a1ex4 replied:
I know I can't, it doesn't pass the tests as it has to be hosted on Pypi. It's a temporary solution while the package is being finished, but thanks for taking the time to review the PR!

a1ex4 committed:
Handle exception during setup, change update function

a1ex4 committed:
Add media_player/androidtv.py to .coveragerc

a1ex4 committed:
Linting

a1ex4 committed:
Updated requirements, linting

a1ex4 committed:
Linting

rytilahti reviewed (commented):
Some general comments about the code, no testing on a real device.

JeffLIrion committed:
Don't access 'self.androidtv._available', get this info from the 'AndroidTV.connect()' and 'AndroidTV.update()' methods

This commit requires an updated `androidtv` package, which is on GitHub but not yet on pypi.  If the code in this component and in the package look good, then I'll push the new version to pypi.

JeffLIrion committed:
Bump androidtv to version 0.0.4

JeffLIrion commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,488 @@
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self.androidtv.muted
```
Yes, they are I/O free.

a1ex4 committed:
Updated setup and updates to work with androidtv 0.0.4

a1ex4 committed:
Merge pull request #2 from JeffLIrion/patch-3

Don't access 'self.androidtv._available', get this info from the 'And…

a1ex4 committed:
Added support for unique_id and device_info

a1ex4 committed:
Better setup check

a1ex4 commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,488 @@
+
+
+ACTIONS = {
+    "back": "4",
```
I would prefer to let it here because as it used for a service, there is a check to see if the requested action exists in the component. Also contributors can more easily add their actions in the component rather than in the backend lib.

a1ex4 commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,488 @@
+}
+
+KNOWN_APPS = {
+    "amazon": "Amazon Prime Video",
```
Same reasons, and also this is very specific to my component (the backend lib is made to work for the FireTV component too).

a1ex4 commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,488 @@
+    add_entities([device])
+    _LOGGER.info("Setup Android TV at %s%s", host, adb_log)
+
+    if host in hass.data[DATA_KEY]:
```
Pushed a commit to better handle that 🙂

a1ex4 commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,488 @@
+
+        self._name = name
+        self._apps = KNOWN_APPS
+        self._apps.update(dict(apps))
```
No, this just means that user can add their custom know apps in the configuration to be added to the KNOWN_APPS dict.

a1ex4 committed:
Fixed the for loop

a1ex4 commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,488 @@
+                app_name = self._apps[app]
+                i += 1
+        if i == 0:
+            app_name = None
```
Yes I'm looking for substring match, thanks for the tip I implemented it 🙂

JeffLIrion committed:
Make 'self.exceptions' a tuple

a1ex4 committed:
Merge pull request #3 from JeffLIrion/patch-4

Make 'self.exceptions' a tuple

JeffLIrion committed:
Catch exceptions in 'adb_wrapper', not 'update'

a1ex4 committed:
Merge pull request #4 from JeffLIrion/patch-5

Catch exceptions in 'adb_wrapper', not 'update'

JeffLIrion committed:
Use 'atv.available' to avoid connecting twice

a1ex4 committed:
Merge pull request #5 from JeffLIrion/patch-6

Use 'atv.available' to avoid connecting twice

arsaboo commented:
Tested on my Shield and got the following error:
```bash
Thu Jan 10 2019 15:03:48 GMT-0500 (Eastern Standard Time)

Error while setting up platform androidtv
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/helpers/entity_platform.py", line 128, in _async_setup_platform
    SLOW_SETUP_MAX_WAIT, loop=hass.loop)
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/homeassistant/.homeassistant/custom_components/media_player/androidtv.py", line 181, in setup_platform
    atv = AndroidTV(host, adbkey)
  File "/srv/homeassistant/lib/python3.6/site-packages/androidtv/__init__.py", line 230, in __init__
    self.properties = self.device_info()
  File "/srv/homeassistant/lib/python3.6/site-packages/androidtv/__init__.py", line 327, in device_info
    btmac = re.findall(BTMAC_REGEX_PATTERN, properties)[0]
IndexError: list index out of range
```

JeffLIrion commented:
I just bumped the version of `androidtv` to 0.0.5 and pushed it to pypi.

bobokun commented:
Also testing on my Shield after changing the requirements of androidtv==0.05 I'm getting an error:


```
Tue Jan 15 2019 15:55:28 GMT-0500 (Eastern Standard Time)

Error while setting up platform androidtv
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/homeassistant/helpers/entity_platform.py", line 128, in _async_setup_platform
    SLOW_SETUP_MAX_WAIT, loop=hass.loop)
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/media_player/androidtv.py", line 181, in setup_platform
    atv = AndroidTV(host, adbkey)
  File "/config/deps/lib/python3.6/site-packages/androidtv/__init__.py", line 230, in __init__
    self.properties = self.device_info()
  File "/config/deps/lib/python3.6/site-packages/androidtv/__init__.py", line 330, in device_info
    wifi_out = self._adb_shell('ip addr show wlan0')
  File "/config/deps/lib/python3.6/site-packages/androidtv/__init__.py", line 474, in _adb_shell_python_adb
    return self._adb.Shell(cmd)
  File "/usr/local/lib/python3.6/site-packages/adb/adb_commands.py", line 376, in Shell
    timeout_ms=timeout_ms)
  File "/usr/local/lib/python3.6/site-packages/adb/adb_protocol.py", line 411, in Command
    return ''.join(cls.StreamingCommand(usb, service, command, timeout_ms))
  File "/usr/local/lib/python3.6/site-packages/adb/adb_protocol.py", line 439, in StreamingCommand
    for data in connection.ReadUntilClose():
AttributeError: 'NoneType' object has no attribute 'ReadUntilClose'
```

JeffLIrion commented:
Reviewers, what's needed to get this merged?

I submitted a couple PRs to @a1ex4 that bump the `androidtv` version to 0.0.5. I think the hold-ups are:

* The `ACTIONS` dictionary
  * I'm not sure how best to handle this. We could instead import it from the `androidtv` package, but I think you're not supposed to import from packages in the main body of the code. But it needs to be in the main body since `ACTIONS` is used in `SERVICE_ACTION_SCHEMA`. The difficult thing is that in order to use that service, the user needs to know the valid actions. 
* The `APPS` dictionary
  * I think this could be moved to the package and imported in the `AndroidTVDevice` class. Maybe all stuff relating to app names should be moved into the package. 
* Setting `self._state = ...`. This could be removed.

Teagan42 commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,524 @@
+        # "pure-python-adb"
+        atv = AndroidTV(
+            host,
+            adb_server_ip=config[CONF_ADB_SERVER_IP],
```
[Question] Shouldn't you accept the adbkey for remote servers in the case someone is hosting ADB  behind auth?

JeffLIrion replied:
The `adbkey` configuration entry is only needed when using the Python implementation of the ADB protocol (aka, the `adb` package).  When using an ADB server for ADB commands (such as the Hass.io addon), the adb instance is responsible for authenticating.  Since that responsibility is outside the scope of this component, the `adbkey` configuration entry is not relevant.

bieniu commented:
Component doeasn't work with HA 0.88b0.
```
Error loading homeassistant.components.media_player.androidtv. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/loader.py", line 147, in _load_file
    module = importlib.import_module(path)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/components/media_player/androidtv.py", line 20, in <module>
    from homeassistant.components.media_player import (
ImportError: cannot import name 'SUPPORT_TURN_OFF'
```

JeffLIrion commented:
> Component doeasn't work with HA 0.88b0.
> 
> ```
> Error loading homeassistant.components.media_player.androidtv. Make sure all dependencies are installed
> Traceback (most recent call last):
>   File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/loader.py", line 147, in _load_file
>     module = importlib.import_module(path)
>   File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 678, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/components/media_player/androidtv.py", line 20, in <module>
>     from homeassistant.components.media_player import (
> ImportError: cannot import name 'SUPPORT_TURN_OFF'
> ```

This pull request should fix the issue: https://github.com/a1ex4/home-assistant/pull/8

In the meantime, this component should work (and be backwards compatible): https://github.com/JeffLIrion/home-assistant/blob/androidtv-backwards-compatible/homeassistant/components/media_player/androidtv.py

arsaboo commented:
@JeffLIrion Does it have to be placed in `media_player/androidtv.py` or `androidtv/media_player`?

JeffLIrion commented:
@arsaboo `media_player/androidtv.py`. If they're changing the directory structure for components in that way, I'm unaware of it.

arsaboo commented:
@JeffLIrion Yes, we are moving to a different folder structure. From the release notes for 0.88
```
Note for custom component developers: We are moving to a new file structure. Platforms now live embedded in components. Custom platforms will have to be updated to follow this pattern. This is a breaking change in case your custom platform overrides a built-in platform. Rename your custom platform from, ie light/hue.py to hue/light.py.
```

If we use it under `media_player/androidtv.py`, I still get the error
```bash
Thu Feb 14 2019 11:57:46 GMT-0500 (Eastern Standard Time)
Error loading custom_components.androidtv.media_player. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/loader.py", line 147, in _load_file
    module = importlib.import_module(path)
  File "/usr/local/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/homeassistant/.homeassistant/custom_components/androidtv/media_player.py", line 33, in <module>
    from homeassistant.components.media_player import (
ImportError: cannot import name 'SUPPORT_TURN_OFF'
``` and a warning about the folder change
```
Log Details (WARNING)
Thu Feb 14 2019 11:57:46 GMT-0500 (Eastern Standard Time)
Integrations need to be in their own folder. Change media_player/androidtv.py to androidtv/media_player.py. This will stop working soon.
```

JeffLIrion commented:
@arsaboo I think it will work if you use the backwards compatible version (and possibly rename it as you mentioned): https://github.com/JeffLIrion/home-assistant/blob/androidtv-backwards-compatible/homeassistant/components/media_player/androidtv.py

tboyce021 commented:
Any chance of including hassio-addons/addon-adb#6 in this?

jjlawren commented:
Looks like the `androidtv` import isn't handling logging properly, dumping every debug message back into HA even when the logger is set to 'warn' with:
```
logger:
    default: warn
```
Every few seconds messages like this are printed in my HA log:
```
03-03-2019:15:35:16,17 DEBUG    [connection.py:24] Connect to adb server - 127.0.0.1:5037
03-03-2019:15:35:16,20 DEBUG    [connection.py:75] b'000Chost:devices'
03-03-2019:15:35:16,22 DEBUG    [connection.py:47] Connection closed...
03-03-2019:15:35:16,23 DEBUG    [connection.py:24] Connect to adb server - 127.0.0.1:5037
03-03-2019:15:35:16,25 DEBUG    [connection.py:24] Connect to adb server - 127.0.0.1:5037
03-03-2019:15:35:16,27 DEBUG    [connection.py:75] b'0027host-serial:10.6.9.12:5555:get-serialno'
03-03-2019:15:35:16,28 DEBUG    [connection.py:47] Connection closed...
03-03-2019:15:35:16,29 DEBUG    [connection.py:24] Connect to adb server - 127.0.0.1:5037
03-03-2019:15:35:16,29 DEBUG    [connection.py:24] Connect to adb server - 127.0.0.1:5037
03-03-2019:15:35:16,30 DEBUG    [connection.py:75] b'001Dhost:transport:10.6.9.12:5555'
03-03-2019:15:35:16,30 DEBUG    [connection.py:75] b"0104shell:dumpsys power | grep 'Display Power' | grep -q 'state=ON' && echo -e '1\\c'  && dumpsys power | grep mWakefulness | grep -q Awake && echo -e '1\\c'  && dumpsys power | grep Locks | grep 'size=' &&dumpsys window windows | grep mCurrentFocus && dumpsys audio"
03-03-2019:15:35:16,51 DEBUG    [connection.py:47] Connection closed...
```

pvizeli commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,524 @@
+    name: MIBOX3
+    adbkey: /config/adbkey
+    apps:
+      "amazon": "Amazon Premium Video"
```
Cleanup this part

pvizeli commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,524 @@
+CONF_ADB_SERVER_IP = 'adb_server_ip'
+CONF_ADB_SERVER_PORT = 'adb_server_port'
+
+DEFAULT_APPS = {}
```
Remove that

pvizeli commented on homeassistant/components/media_player/androidtv.py:
```diff
@@ -0,0 +1,524 @@
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): vol.All(cv.port, cv.string),
+    vol.Optional(CONF_ADBKEY): has_adb_files,
+    vol.Optional(CONF_APPS, default=DEFAULT_APPS): dict,
```
`default=dict()`

pvizeli reviewed (commented):
Small comments, after that we can merge it

JeffLIrion commented:
@pvizeli I have a newer PR that combines the `androidtv` and `firetv` components (for ease of maintenance). Can we merge that PR instead. 

https://github.com/home-assistant/home-assistant/pull/21944

cgtobi commented:
Since #21944 has been merged we can close this, no?
